trace/collector.go

```diff
@@ -0,0 +1,49 @@
+package trace
+
+import (
+	"log"
+	"sync"
+)
+
+type Collector interface {
+	Record(e Event) error
+}
+
+type MemoryCollector struct {
+	list []Event
+	in   chan Event
+	lock sync.Mutex
+}
+
+func (c MemoryCollector) Record(event Event) error {
+	c.in <- event
+	return nil
+}
+
+func (c MemoryCollector) GetEvents() []Event {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	listLength := len(c.list)
+	newList := make([]Event, listLength)
+	copied := copy(newList, c.list)
+	if copied != listLength {
+		log.Panicln("Failed to copy Events from list")
+	}
+	return newList
+}
+
+func NewMemoryCollector() *MemoryCollector {
+	c := MemoryCollector{}
+	c.list = []Event{}
```

Comment:
You might want to initialize this slice with a biggish capacity so the copying can be minimized.

```go
const NumberOfEventsBeforeFlush = 1000

...

c.list = make([]Event, 0, NumberOfEventsBeforeFlush)
```